### PR TITLE
Suggest `try_into` when possible

### PR DESCRIPTION
**Pull request by estebank**
CC #47168

### CODE REVIEW

File: src/test/ui/integer-literal-suffix-inference.stderr [hunk 1]
```diff
@@ -3,288 +3,342 @@ error[E0308]: mismatched types
...
    |
 LL |     id_i16(c8);
-   |            ^^ expected i16, found i8
-help: you can cast an `i8` to `i16`, which will sign-extend the source value
-   |
-LL |     id_i16(c8.into());
-   |            ^^^^^^^^^
+   |            ^^
+   |            |
+   |            expected i16, found i8
```
Comment (oli-obk) [inline src/test/ui/integer-literal-suffix-inference.stderr]: I wonder if we should change the `mismathced types` root message to `expected i16, found i8` so the suggestion can be inlined properly

Comment (estebank) [reply src/test/ui/integer-literal-suffix-inference.stderr]: It took me a second to understand but now I see what you mean... That's probably something worthy of doing in a separate PR. The actual code change should be smallish, but the effect in `stderr` files will probably be extensive.

File: src/test/ui/repeat_count.stderr [hunk 2]
```diff
@@ -42,12 +42,20 @@ error[E0308]: mismatched types
...
+   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error[E0308]: mismatched types
   --> $DIR/repeat_count.rs:28:23
    |
 LL |     let f = [0_usize; -1_isize];
    |                       ^^^^^^^^ expected usize, found isize
+help: you can convert an `isize` to `usize` or panic if it the converted value wouldn't fit
+   |
+LL |     let f = [0_usize; (-1_isize).try_into().unwrap()];
```
Comment (oli-obk) [inline src/test/ui/repeat_count.stderr]: for now ok, but we should try to not suggest code that is known to panic

**Review by oli-obk — COMMENTED** [review]
how does this code interact with constants? Do we have some tests for that? For now we should not be suggesting that `const FOO: u32 = bar();` should use `try_from`, or `from` irrelevant of the return type of `bar`

File: src/test/ui/repeat_count.stderr (hunk 2 above)
Comment (estebank) [reply src/test/ui/repeat_count.stderr]: This is indeed a bad corner case :-/

**Comment by estebank** [comment]
> how does this code interact with constants? 

I'm checking but, I think we have any guards against that at the moment, given [this current example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b253efa3ddd941fd3e564786e884fac1) where rustc gives suggestions in expr context, but not const context.

File: src/test/ui/repeat_count.stderr (hunk 2 above)
Comment (estebank) [reply src/test/ui/repeat_count.stderr]: Filed https://github.com/rust-lang/rust/issues/60384

**Comment by oli-obk** [comment]
You'll also need to extend that check to know about `const fn`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=03af30e0c23632bd5b03bcde9385b116

**Comment by estebank** [comment]
@oli-obk oh :(

**Comment by estebank** [comment]
@oli-obk `const fn`s are now accounted for.

**Pull request merged**